### ISSUE TITLE
:initial state doesn't work in Rails 4

### ISSUE THREAD

**Issue by whatthewhat** [pluginaweek/state_machine#235]
With current rails 4 master initial state doesn't get saved to db:

With this state_machine defenition:

``` ruby
class Article < ActiveRecord::Base
  state_machine :initial => :draft do
    event :publish do
      transition :draft => :published
    end
  end
end
```

I get this:

``` ruby
article = Article.new
# => <Article id: nil, created_at: nil, updated_at: nil, state: "draft">
article.state
# => "draft"

# but
article.changes
# => {}
article.save
# => true
article.reload.state
# => nil
```

So the state value doen't get written to the database.

**Comment by obrie** [comment]
Good catch!  This only happens when there are no changes to record for a new record (i.e. it uses all the default values in the db).  The fix isn't immediately clear to me so this may take just a little while to figure out what the right solution is.

**Comment by obrie** [comment]
This is being tracked over at https://github.com/rails/rails/pull/9489 .. if that PR gets merged, everything will start working again in Rails 4

**Comment by obrie** [comment]
In anyone is interested in Rails 4 support, please jump over to that bug and voice your support -- this will be blocked until the issue is addressed in Rails.

**Comment by simonmd** [comment]
In the meantime, what would be the most sensible workaround for this?

**Comment by the8472** [comment]
as @obrie said

> This only happens when there are no changes to record for a new record

So.. touching some value, e.g. manually setting `updated_at` might work.

**Comment by afinetooth** [comment]
FYI, just noticed the the pull request was merged:
https://github.com/rails/rails/pull/9489

**Comment by bai** [comment]
The issue doesn't seem to be fixed with that rails commit. Is anyone digging into it?

**Comment by obrie** [comment]
It looks like this will still be a problem if you're using the `protected_attributes` gem -- it seems to unfortunately completely reimplement the `initialize` method.  See https://github.com/rails/protected_attributes/blob/master/lib/active_record/mass_assignment_security/core.rb

@bai can you verify that this is what you're seeing?

**Comment by legolin** [comment]
Would hooking into that :initialize callback work, or would that cause other issues?

**Comment by obrie** [comment]
@legolin it causes other issues.  The current implementation is pretty well-supported.  The protected_attributes gem is just doing something that's not maintainable -- they have to constantly be sync'ing with the latest implementation of the `initialize` method in ActiveRecord.  Someone (could be myself) will need to bring this up to whoever's maintaining that project.

**Comment by obrie** [comment]
Also that reimplementation doesn't seem to be doing anything interesting -- not sure why it's there in the first place.

**Comment by obrie** [comment]
If you're using `protected_attributes`, https://github.com/rails/protected_attributes/pull/14 will need to get merged and deployed in order for `state_machine` to completely fix the issue

**Comment by obrie** [comment]
Going to close this since state_machine now works in the default Rails 4 installation.. the remaining broken library (protected_attributes) will be tracked in that project's repo.

**Comment by scarver2** [comment]
This issue is prevalent in Rails 4.0.0.rc1. Initial state is defaulting to nil. The only way to overcome this issue was to write a migration to set the initial state.

`class SetInitialStateOfThing < ActiveRecord::Migration
  def change
    change_column :things, :state, :string, default: 'initial_state', null: false    
  end
end`

**Comment by obrie** [comment]
@scarver2 Are you using the `protected_attributes` gem?  No one has yet to merge the PR required to prevent `protected_attributes` from breaking `state_machine` -- see https://github.com/rails/protected_attributes/pull/14

**Comment by scarver2** [comment]
Yes, I'm using the `protected_attributes` gem. The migration I added is actually a best practice to guard against junk in the database. I feel better having the defaults in place and being able to continue Rails 4 testing.  

Speaking of gems, "Thank You" for the `state_machine` gem.

**Comment by simonmorley** [comment]
Why didn't I use Google before wasting a morning on this. Bangs head hard against wall...

**Comment by obrie** [comment]
@simonmorley Sorry you wasted time on this! :/  Hopefully they release a new version soon...

**Comment by simonmorley** [comment]
This actually helped me no end. Previously I'd got about 3 tests that I simply could not get to pass since the upgrade to rc1. No matter what I tried, the update action wouldn't actually save the values for some fields in some scenarios. Annoying. And no errors. Seriously thought I was going crazy.

In the end, using protected_attributes from git directly in my gemfile helped but still saw odd results. Don't have time to find out why right now - ultimately I ditched it and converted the whole app to strong attr. Wasnt so bad with tests already in place.
